homeassistant/components/hue/sensor_base.py

```diff
@@ -0,0 +1,273 @@
+        self._started = True
+        _LOGGER.info('Starting sensor polling loop with %s second interval',
+                     self.SCAN_INTERVAL.total_seconds())
+
+        async def async_update_bridge(now):
+            """Will update sensors from the bridge."""
+            await self.async_update_items()
+
+            async_track_point_in_utc_time(
+                self.hass, async_update_bridge, utcnow() + self.SCAN_INTERVAL)
+
+        await async_update_bridge(None)
+
+    async def async_update_items(self):
+        """Update sensors from the bridge."""
+        import aiohue
+
+        api = self.bridge.api.sensors
+
+        try:
+            start = monotonic()
+            with async_timeout.timeout(4):
+                await api.update()
+        except (asyncio.TimeoutError, aiohue.AiohueException) as err:
+            _LOGGER.debug('Failed to fetch sensor: %s', err)
+
+            if not self.bridge.available:
+                return
+
+            _LOGGER.error('Unable to reach bridge %s (%s)', self.bridge.host,
+                          err)
+            self.bridge.available = False
+
+            return
+
+        finally:
+            _LOGGER.debug('Finished sensor request in %.3f seconds',
+                          monotonic() - start)
+
+        if not self.bridge.available:
+            _LOGGER.info('Reconnected to bridge %s', self.bridge.host)
+            self.bridge.available = True
+
+        new_sensors = []
+        new_binary_sensors = []
+        primary_sensor_devices = {}
+        current = self.hass.data[hue.DOMAIN][CURRENT_SENSORS]
+
+        # Physical Hue motion sensors present as three sensors in the API: a
+        # presence sensor, a temperature sensor, and a light level sensor. Of
+        # these, only the presence sensor is assigned the user-friendly name
+        # that the user has given to the device. Each of these sensors is
+        # linked by a common device_id, which is the first twenty-three
+        # characters of the unique id (then followed by a hyphen and an ID
+        # specific to the individual sensor).
+        #
+        # To set up neat values, and assign the sensor entities to the same
+        # device, we first, iterate over all the sensors and find the Hue
+        # presence sensors, then iterate over all the remaining sensors -
+        # finding the remaining ones that may or may not be related to the
+        # presence sensors.
+        for item_id in api:
+            if api[item_id].type != aiohue.sensors.TYPE_ZLL_PRESENCE:
+                continue
+
+            primary_sensor_devices[_device_id(api[item_id])] = api[item_id]
+
+        # Iterate again now we have all the presence sensors, and add the
+        # related sensors with nice names where appropriate.
+        for item_id in api:
+            existing = current.get(api[item_id].uniqueid)
+            if existing is not None:
+                self.hass.async_create_task(existing.async_update_ha_state())
```

Comment:
This doesn't guarantee that the entity has been added to home assistant. We use our dispatch helper for this case.

Reply:
Could you give an example? I'm having trouble finding one in the codebase or documentation?

Reply:
https://github.com/home-assistant/home-assistant/blob/7862fdd27eafb094f38b697feb761f1c29e9e29f/homeassistant/components/mysensors/device.py#L131-L137

https://github.com/home-assistant/home-assistant/blob/38d92b2abfa5a4ced530ecb809fa5e7c1ead568d/homeassistant/components/mysensors/handler.py#L88-L102

Reply:
I don't think that applies here. This code is part of a loop that runs every `SCAN_INTERVAL`, it doesn't need to be called when the entity is first added as it already has all the state it needs; its purpose is solely to update those entities that already exist.

Reply:
We will call `async_add_entities` each time there are new entities, correct? Then we can't guarantee that `async_add_entities` has finished before another update comes in and we want to call update on the entities. When `async_setup_entry` returns we only wait for the entities that are in the process of being added are done. Entities that are added later doesn't have this guarantee, so we risk calling update before they are added to home assistant, which will give an error. That's why we need to use the dispatcher.

Reply:
> we risk calling update before they are added to home assistant, which will give an error

But do we actually care about that error? I'd rather just ignore it and let `async_setup_entry` finish its thing, than try to compensate for it and overcomplicate this. We're talking about a situation that will occur only if scanning the bridge and adding all the new entities takes longer than 5 seconds. Sure it's possible, but it's not going to be common, and the impact of ignoring the error is negligible (since `async_setup_entry` will finish and set the state as soon as it can anyway).

Reply:
Added some error handling with the last push.